### PR TITLE
Checkstyle: fix NeedBraces properties #319

### PR DESCRIPTION
**Pull request by eugenepeh**
#319
Ready for review

### CODE REVIEW

**Comment by CanIHasReview** [comment]
# v1

@eugenepeh submitted v1 for review.

* [v1 1/2] [Checkstyle: fix NeedBraces properties](https://github.com/se-edu/addressbook-level4/pull/362/commits/f28e67f7042d5e246693776befe11576e1c4b6d1)
* [v1 2/2] [Checkstyle: fix NeedBraces rule violations ](https://github.com/se-edu/addressbook-level4/pull/362/commits/0e2684a523b276d2e1ba5247a3b43e0bb77bab9e)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/27cb4621dca1bf474b2838352a90873f3fbcb4ef...0e2684a523b276d2e1ba5247a3b43e0bb77bab9e))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/362/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by eugenepeh** [comment]
Thanks @yamgent, the "for debugging purposes" is quoted from the oss java coding stand 
(https://oss-generic.github.io/process/codingStandards/CodingStandard-Java.html),
it is to helps the reader to understand why this fix is necessary.

Should I still remove it from commit message? or should I elaborate further?

**Comment by yamgent** [comment]
@eugenepeh Ah I see, thanks for pointing that out. You can leave them in (in fact, you can combine the first and second paragraph since they both establish the context).

Actually, perhaps the sentence can be made clearer then, because it took me some time to understand what was going on.

> it is not apparent whether the test is really true or not.

`it is not apparent whether the **conditional clause evaluates to** true or **false**.`

**Comment by CanIHasReview** [comment]
# v2

@eugenepeh submitted v2 for review.

* [v2 1/1] [Checkstyle: fix NeedBraces properties and rule violations](https://github.com/se-edu/addressbook-level4/pull/362/commits/4cc7c2ef6bc08d163a40b1a062e3a04f0def6ecb)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/27cb4621dca1bf474b2838352a90873f3fbcb4ef...4cc7c2ef6bc08d163a40b1a062e3a04f0def6ecb)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/compare/e132a02dfcc5a89a3a6e3d9d850a42469421a462~1...e132a02dfcc5a89a3a6e3d9d850a42469421a462))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/362/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by CanIHasReview** [comment]
# v3

@eugenepeh submitted v3 for review.

* [v3 1/1] [Checkstyle: fix NeedBraces properties](https://github.com/se-edu/addressbook-level4/pull/362/commits/385231d03fe0e7dbad3a8f98f39a870397fc37e4)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/27cb4621dca1bf474b2838352a90873f3fbcb4ef...385231d03fe0e7dbad3a8f98f39a870397fc37e4)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/compare/c76ba14009a3b00972ac1ae422ecaac105150615~1...c76ba14009a3b00972ac1ae422ecaac105150615))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/362/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: config/checkstyle/checkstyle.xml [hunk 1]
```diff
@@ -204,7 +204,7 @@
         return 1; // Not allowed
       -->
       <property name="allowEmptyLoopBody" value="true"/>
-      <property name="allowSingleLineStatement" value="true"/>
+      <property name="allowSingleLineStatement" value="false"/>
```
Comment (PierceAndy) [inline config/checkstyle/checkstyle.xml]: Default value. Line can be omitted.

**Comment by CanIHasReview** [comment]
# v4

@eugenepeh submitted v4 for review.

* [v4 1/1] [Checkstyle: fix NeedBraces properties](https://github.com/se-edu/addressbook-level4/pull/362/commits/58ac2423d38b633a1c1fdd230aa24938dafaae49)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/a61db1bf27f62cc549eb79be0a4598f2f1994b9b...58ac2423d38b633a1c1fdd230aa24938dafaae49)) ([:chart_with_upwards_trend: Interdiff between v3 and v4](https://github.com/se-edu/addressbook-level4/compare/cc53913c95cd66417735a0e186932bdf5b133c27~1...cc53913c95cd66417735a0e186932bdf5b133c27))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/362/4/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by CanIHasReview** [comment]
# v5

@eugenepeh submitted v5 for review.

* [v5 1/1] [Checkstyle: fix NeedBraces properties](https://github.com/se-edu/addressbook-level4/pull/362/commits/56fd00f54e250f75128fb26e3d8bad4b52ede927)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/07367fab5054768f3243fb55eba1367d3d67a39b...56fd00f54e250f75128fb26e3d8bad4b52ede927)) ([:chart_with_upwards_trend: Interdiff between v4 and v5](https://github.com/se-edu/addressbook-level4/compare/f3ae15831ca2d048d6f82c4a65cfdc9217509ac6~1...f3ae15831ca2d048d6f82c4a65cfdc9217509ac6))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/362/5/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by damithc** [comment]
This needs a bit more deliberation:
1. I think the problem with single line conditional is not related to debugging. A good IDE should be able to handle this case. The problem is the potential to introduce bugs like this.
```java
if(something) foo();
    bar();
```
So the commit message needs to be updated.

2. Should we allow this `if(something) { foo(); }` ?. i.e. we make the braces compulsory, but allow this compact form. It saves 2 extra lines, especially useful when used for guard clauses.

**Comment by PierceAndy** [comment]
With regards to point 2:
We could, since the braces make the scope quite explicit. But doing so means that this form would be allowed for methods as well. 

Personally, I'm not a fan of this style, as Java tends to be verbose, and these statements tend to end up being quite long, and thus hard to read.

**Comment by eugenepeh** [comment]
> Should we allow this if(something) { foo(); } ?. i.e. we make the braces compulsory, but allow this compact form. It saves 2 extra lines, especially useful when used for guard clauses.

Another problem with allowing single line conditional is that, it seems to cause test coverage checker such as coverall to produce inaccurate coverage. Hence the coverage decreased for this PR.

Should we still allow it this way?

**Comment by damithc** [comment]
I'm ok to disallow single line conditionals even with the braces.
@eugenepeh can update the commit message to reflect the rationale (i.e. point 1) ?

**Comment by damithc** [comment]
Revised the PR title and commit message a bit.

**Pull request merged**